lib/DxcSupport/dxcmem.cpp

```diff
@@ -98,3 +98,39 @@ DxcThreadMalloc::DxcThreadMalloc(IMalloc *pMallocOrNull) throw() {
 DxcThreadMalloc::~DxcThreadMalloc() {
     DxcSwapThreadMalloc(pPrior, nullptr);
 }
+
+void* DxcNew(std::size_t size) {
+  void *ptr;
+  IMalloc* iMalloc = DxcGetThreadMallocNoRef();
+  if (iMalloc != nullptr) {
+    ptr = iMalloc->Alloc(size);
+  } else {
+    // DxcGetThreadMallocNoRef() returning null means the operator is called before DllMain
+    // where the g_pDefaultMalloc is initialized, for example from CRT libraries when
+    // static linking is enabled. In that case fallback to the standard allocator
+    // and use CoTaskMemAlloc directly instead of CoGetMalloc, Alloc & Release for better perf.
+#if defined(_WIN32) && !defined(DXC_DISABLE_ALLOCATOR_OVERRIDES)
+    ptr = CoTaskMemAlloc(size);
+#else
+    ptr = malloc(size);
+#endif
+  }
+  return ptr;
+}
+
+void DxcDelete(void *ptr) {
```

Comment:
Is this safe? If an allocation gets allocated before `DllMain` and de-allocated after a custom allocator is enabled, what happens? Does that route the free to the custom allocator?

Reply:
Alloc/free should always happen in the same context. The custom allocator is only enabled inside of compiler API calls. Every public API implementation uses `DxcThreadMalloc` local var to make sure the custom allocator is set to the thread-local storage for use and then re-set back to the previous one on exit.
@tex3d  - please correct me if I am wrong: The situation where something gets allocated by the default allocator (before or after `DllMain`) and de-allocated with a custom allocator should not happen.